libs/langchain/langchain/chains/api/base.py

```diff
@@ -69,7 +79,111 @@ class APIChain(Chain):
             what network access it has.
 
             See https://python.langchain.com/docs/security for more information.
-        """
+
+        Note: this class is deprecated. See below for a replacement implementation
+        using LangGraph. The benefits of this implementation are:
+
+        - Uses LLM tool calling features to encourage properly-formatted API requests;
+        - Support for both token-by-token and step-by-step streaming;
+        - Support for checkpointing and memory of chat history;
+        - Easier to modify or extend (e.g., with additional tools, structured responses, etc.)
+
+            .. code-block:: python
+
+                from typing import Annotated, Sequence
+                from typing_extensions import TypedDict
+
+                from langchain.chains.api.prompt import API_URL_PROMPT
+                from langchain_community.agent_toolkits.openapi.toolkit import RequestsToolkit
+                from langchain_community.utilities.requests import TextRequestsWrapper
+                from langchain_core.messages import BaseMessage
+                from langchain_core.prompts import ChatPromptTemplate
+                from langchain_openai import ChatOpenAI
+                from langchain_core.runnables import RunnableConfig
+                from langgraph.graph import END, StateGraph
+                from langgraph.graph.message import add_messages
+                from langgraph.prebuilt.tool_node import ToolNode
+
+                # NOTE: There are inherent risks in giving models discretion
+                # to execute real-world actions. We must "opt-in" to these
+                # risks by setting allow_dangerous_request=True to use these tools.
+                # This can be dangerous for calling unwanted requests. Please make
+                # sure your custom OpenAPI spec (yaml) is safe and that permissions
+                # associated with the tools are narrowly-scoped.
+                ALLOW_DANGEROUS_REQUESTS = True
+
+                # Subset of spec for https://jsonplaceholder.typicode.com
+                api_spec = \"\"\"
```

Comment:
does it matter for yaml that these are tabbed

Reply:
fixed, ty